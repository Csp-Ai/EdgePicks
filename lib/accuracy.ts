import { getSupabaseClient } from './supabaseClient';
import { agents as agentRegistry } from './agents/registry';
import type { AgentName, AgentOutputs } from './types';

interface MatchupRow {
  agents: AgentOutputs;
  pick: { winner?: string } | null;
  actual_winner: string | null;
  flow: string | null;
}

export interface AccuracyStat {
  name: string;
  wins: number;
  losses: number;
  accuracy: number;
}

export async function recomputeAccuracy() {
  const client = getSupabaseClient();
  const { data, error } = await client
    .from('matchups')
    .select('agents, pick, actual_winner, flow');
  if (error || !data) {
    throw new Error(error?.message || 'Failed to fetch matchups');
  }

  const agentTallies: Record<AgentName, { wins: number; losses: number }> = {};
  agentRegistry.forEach(({ name }) => {
    agentTallies[name as AgentName] = { wins: 0, losses: 0 };
  });
  const flowTallies: Record<string, { wins: number; losses: number }> = {};

  data.forEach((row: MatchupRow) => {
    const actual = row.actual_winner;
    if (!actual) return;

    // Agent stats
    agentRegistry.forEach(({ name }) => {
      const pick = row.agents?.[name]?.team;
      if (!pick) return;
      if (pick === actual) {
        agentTallies[name as AgentName].wins += 1;
      } else {
        agentTallies[name as AgentName].losses += 1;
      }
    });

    // Flow stats
    const flowName = row.flow || 'unknown';
    const predicted = row.pick?.winner;
    if (!flowTallies[flowName]) flowTallies[flowName] = { wins: 0, losses: 0 };
    if (predicted) {
      if (predicted === actual) {
        flowTallies[flowName].wins += 1;
      } else {
        flowTallies[flowName].losses += 1;
      }
    }
  });

  const agentStats = Object.entries(agentTallies).map(([agent, t]) => ({
    agent,
    wins: t.wins,
    losses: t.losses,
    accuracy: t.wins + t.losses > 0 ? t.wins / (t.wins + t.losses) : 0,
  }));

  const flowStats = Object.entries(flowTallies).map(([flow, t]) => ({
    flow,
    wins: t.wins,
    losses: t.losses,
    accuracy: t.wins + t.losses > 0 ? t.wins / (t.wins + t.losses) : 0,
  }));

  await client.from('agent_stats').upsert(agentStats, { onConflict: 'agent' });
  await client.from('flow_stats').upsert(flowStats, { onConflict: 'flow' });

  return { agentStats, flowStats };
}

