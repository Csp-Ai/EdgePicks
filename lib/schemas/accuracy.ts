import { z } from 'zod';

const accuracyEntrySchema = z.object({
  name: z.string(),
  wins: z.number(),
  losses: z.number(),
  accuracy: z.number(),
});

export const accuracySchema = z.object({
  agents: z.array(accuracyEntrySchema),
  flows: z.array(accuracyEntrySchema),
});

export type AccuracyResponse = z.infer<typeof accuracySchema>;
