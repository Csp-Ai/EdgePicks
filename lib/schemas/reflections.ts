import { z } from 'zod';

export const agentReflectionSchema = z.object({
  whatIObserved: z.string(),
  whatIChose: z.string(),
  whatCouldImprove: z.string(),
});

export const reflectionsSchema = z.record(agentReflectionSchema);

export type ReflectionsResponse = z.infer<typeof reflectionsSchema>;
