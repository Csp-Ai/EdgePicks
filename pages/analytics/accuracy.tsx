import { GetServerSideProps } from 'next';
import AccuracyTrend, { AccuracyHistory } from '../../components/AccuracyTrend';

interface Props {
  fallbackData: AccuracyHistory;
}

const AccuracyAnalyticsPage: React.FC<Props> = ({ fallbackData }) => (
  <main className="min-h-screen p-6 bg-gray-50" suppressHydrationWarning>
    <header className="text-center mb-8">
      <h1 className="text-3xl font-mono font-bold">Accuracy Trends</h1>
      <p className="text-gray-600">Weekly accuracy per agent and overall</p>
    </header>
    <AccuracyTrend fallbackData={fallbackData} />
  </main>
);

export const getServerSideProps: GetServerSideProps<Props> = async ({ req }) => {
  const baseUrl = req.headers.host ? `http://${req.headers.host}` : '';
  try {
    const res = await fetch(`${baseUrl}/api/accuracy-history`);
    const fallbackData: AccuracyHistory = await res.json();
    return { props: { fallbackData } };
  } catch {
    return { props: { fallbackData: { history: [] } } };
  }
};

export default AccuracyAnalyticsPage;
