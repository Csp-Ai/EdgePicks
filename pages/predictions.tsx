import React from 'react';
import MatchupInsights from '@/components/predictions/MatchupInsights';

const PredictionsPage: React.FC = () => {
  return <MatchupInsights />;
};

export default PredictionsPage;
