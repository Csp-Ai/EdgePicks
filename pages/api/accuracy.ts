import type { NextApiRequest, NextApiResponse } from 'next';
import { supabase } from '@/lib/supabaseClient';
import { recomputeAccuracy } from '@/lib/accuracy';

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  if (req.method === 'POST') {
    const { matchupId, actualWinner } = req.body as {
      matchupId?: string;
      actualWinner?: string;
    };
    if (!matchupId || !actualWinner) {
      res.status(400).json({ error: 'matchupId and actualWinner are required' });
      return;
    }

    const { error } = await supabase
      .from('matchups')
      .update({ actual_winner: actualWinner })
      .eq('id', matchupId);
    if (error) {
      res.status(500).json({ error: error.message });
      return;
    }

    try {
      await recomputeAccuracy();
    } catch (err: any) {
      console.error('recomputeAccuracy failed', err);
    }

    res.status(200).json({ success: true });
    return;
  }

  if (req.method === 'GET') {
    const { data: agentRows, error: agentError } = await supabase
      .from('agent_stats')
      .select('*')
      .order('accuracy', { ascending: false });
    const { data: flowRows, error: flowError } = await supabase
      .from('flow_stats')
      .select('*')
      .order('accuracy', { ascending: false });
    if (agentError || flowError) {
      res
        .status(500)
        .json({ error: agentError?.message || flowError?.message });
      return;
    }

    res.status(200).json({
      agents: (agentRows || []).map((r) => ({
        name: r.agent,
        wins: r.wins,
        losses: r.losses,
        accuracy: r.accuracy,
      })),
      flows: (flowRows || []).map((r) => ({
        name: r.flow,
        wins: r.wins,
        losses: r.losses,
        accuracy: r.accuracy,
      })),
    });
    return;
  }

  res.setHeader('Allow', ['GET', 'POST']);
  res.status(405).end('Method Not Allowed');
}

