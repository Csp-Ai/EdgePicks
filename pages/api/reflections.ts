import type { NextApiRequest, NextApiResponse } from 'next';
import { readAgentReflections } from '../../lib/writeAgentReflection';

export default async function handler(_req: NextApiRequest, res: NextApiResponse) {
  const data = await readAgentReflections();
  res.status(200).json(data);
}
