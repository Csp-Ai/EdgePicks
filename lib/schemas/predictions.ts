import { z } from 'zod';
import { agentReflectionSchema } from './reflections';

const agentResultSchema = z.object({
  team: z.string(),
  score: z.number(),
  reason: z.string().optional(),
  warnings: z.array(z.string()).optional(),
  reflection: agentReflectionSchema.optional(),
  weight: z.number().optional(),
  scoreTotal: z.number().optional(),
  confidenceEstimate: z.number().optional(),
  description: z.string().optional(),
});

const agentExecutionSchema = z.object({
  name: z.string(),
  result: agentResultSchema.optional(),
  error: z.literal(true).optional(),
  errorInfo: z
    .object({ message: z.string().optional(), stack: z.string().optional() })
    .optional(),
  scoreTotal: z.number().optional(),
  confidenceEstimate: z.number().optional(),
  agentDurationMs: z.number().optional(),
  sessionId: z.string().optional(),
});

const gameSchema = z.object({
  homeTeam: z.object({ name: z.string() }),
  awayTeam: z.object({ name: z.string() }),
  time: z.string(),
});

const predictionSchema = z.object({
  game: gameSchema,
  winner: z.string(),
  confidence: z.number(),
  agents: z.record(agentResultSchema),
  agentScores: z.record(z.number()),
  executions: z.array(agentExecutionSchema),
});

export const predictionsSchema = z.object({
  predictions: z.array(predictionSchema),
  agentScores: z.record(z.number()),
  weightsUsed: z.record(z.number()),
  timestamp: z.string(),
  cacheVersion: z.string(),
});

export type PredictionsResponse = z.infer<typeof predictionsSchema>;
