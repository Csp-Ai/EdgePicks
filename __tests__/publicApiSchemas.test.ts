/** @jest-environment node */
import { zodToJsonSchema } from 'zod-to-json-schema';
import { accuracySchema } from '../lib/schemas/accuracy';
import { reflectionsSchema } from '../lib/schemas/reflections';
import { predictionsSchema } from '../lib/schemas/predictions';

describe('public API schema snapshots', () => {
  test('accuracy schema', () => {
    const schema = zodToJsonSchema(accuracySchema, 'Accuracy');
    expect(schema).toMatchSnapshot();
  });

  test('reflections schema', () => {
    const schema = zodToJsonSchema(reflectionsSchema, 'Reflections');
    expect(schema).toMatchSnapshot();
  });

  test('predictions schema', () => {
    const schema = zodToJsonSchema(predictionsSchema, 'Predictions');
    expect(schema).toMatchSnapshot();
  });
});
