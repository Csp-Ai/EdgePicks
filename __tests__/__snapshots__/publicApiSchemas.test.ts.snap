// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`public API schema snapshots accuracy schema 1`] = `
{
  "$ref": "#/definitions/Accuracy",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "Accuracy": {
      "additionalProperties": false,
      "properties": {
        "agents": {
          "items": {
            "additionalProperties": false,
            "properties": {
              "accuracy": {
                "type": "number",
              },
              "losses": {
                "type": "number",
              },
              "name": {
                "type": "string",
              },
              "wins": {
                "type": "number",
              },
            },
            "required": [
              "name",
              "wins",
              "losses",
              "accuracy",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "flows": {
          "items": {
            "$ref": "#/definitions/Accuracy/properties/agents/items",
          },
          "type": "array",
        },
      },
      "required": [
        "agents",
        "flows",
      ],
      "type": "object",
    },
  },
}
`;

exports[`public API schema snapshots predictions schema 1`] = `
{
  "$ref": "#/definitions/Predictions",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "Predictions": {
      "additionalProperties": false,
      "properties": {
        "agentScores": {
          "additionalProperties": {
            "type": "number",
          },
          "type": "object",
        },
        "cacheVersion": {
          "type": "string",
        },
        "predictions": {
          "items": {
            "additionalProperties": false,
            "properties": {
              "agentScores": {
                "additionalProperties": {
                  "type": "number",
                },
                "type": "object",
              },
              "agents": {
                "additionalProperties": {
                  "additionalProperties": false,
                  "properties": {
                    "confidenceEstimate": {
                      "type": "number",
                    },
                    "description": {
                      "type": "string",
                    },
                    "reason": {
                      "type": "string",
                    },
                    "reflection": {
                      "additionalProperties": false,
                      "properties": {
                        "whatCouldImprove": {
                          "type": "string",
                        },
                        "whatIChose": {
                          "type": "string",
                        },
                        "whatIObserved": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "whatIObserved",
                        "whatIChose",
                        "whatCouldImprove",
                      ],
                      "type": "object",
                    },
                    "score": {
                      "type": "number",
                    },
                    "scoreTotal": {
                      "type": "number",
                    },
                    "team": {
                      "type": "string",
                    },
                    "warnings": {
                      "items": {
                        "type": "string",
                      },
                      "type": "array",
                    },
                    "weight": {
                      "type": "number",
                    },
                  },
                  "required": [
                    "team",
                    "score",
                  ],
                  "type": "object",
                },
                "type": "object",
              },
              "confidence": {
                "type": "number",
              },
              "executions": {
                "items": {
                  "additionalProperties": false,
                  "properties": {
                    "agentDurationMs": {
                      "type": "number",
                    },
                    "confidenceEstimate": {
                      "type": "number",
                    },
                    "error": {
                      "const": true,
                      "type": "boolean",
                    },
                    "errorInfo": {
                      "additionalProperties": false,
                      "properties": {
                        "message": {
                          "type": "string",
                        },
                        "stack": {
                          "type": "string",
                        },
                      },
                      "type": "object",
                    },
                    "name": {
                      "type": "string",
                    },
                    "result": {
                      "$ref": "#/definitions/Predictions/properties/predictions/items/properties/agents/additionalProperties",
                    },
                    "scoreTotal": {
                      "type": "number",
                    },
                    "sessionId": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "name",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
              "game": {
                "additionalProperties": false,
                "properties": {
                  "awayTeam": {
                    "additionalProperties": false,
                    "properties": {
                      "name": {
                        "type": "string",
                      },
                    },
                    "required": [
                      "name",
                    ],
                    "type": "object",
                  },
                  "homeTeam": {
                    "additionalProperties": false,
                    "properties": {
                      "name": {
                        "type": "string",
                      },
                    },
                    "required": [
                      "name",
                    ],
                    "type": "object",
                  },
                  "time": {
                    "type": "string",
                  },
                },
                "required": [
                  "homeTeam",
                  "awayTeam",
                  "time",
                ],
                "type": "object",
              },
              "winner": {
                "type": "string",
              },
            },
            "required": [
              "game",
              "winner",
              "confidence",
              "agents",
              "agentScores",
              "executions",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "timestamp": {
          "type": "string",
        },
        "weightsUsed": {
          "additionalProperties": {
            "type": "number",
          },
          "type": "object",
        },
      },
      "required": [
        "predictions",
        "agentScores",
        "weightsUsed",
        "timestamp",
        "cacheVersion",
      ],
      "type": "object",
    },
  },
}
`;

exports[`public API schema snapshots reflections schema 1`] = `
{
  "$ref": "#/definitions/Reflections",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "Reflections": {
      "additionalProperties": {
        "additionalProperties": false,
        "properties": {
          "whatCouldImprove": {
            "type": "string",
          },
          "whatIChose": {
            "type": "string",
          },
          "whatIObserved": {
            "type": "string",
          },
        },
        "required": [
          "whatIObserved",
          "whatIChose",
          "whatCouldImprove",
        ],
        "type": "object",
      },
      "type": "object",
    },
  },
}
`;
